gamestate/gamestate.py

```diff
@@ -1,278 +1,282 @@
+        for team in ['blue', 'yellow']:
+            for robot_id in self.get_robot_ids(team):
+                robot_pos = self.get_robot_position(team, robot_id)
+                if robot_pos is not None:
+                    key = (team, robot_id)
+                    all_robot_positions.append((key, robot_pos))
+        return all_robot_positions
+
+    def update_robot_position(self, team, robot_id, pos):
+        assert(len(pos) == 3 and type(pos) == np.ndarray)
+        pos = pos.copy().astype(float)
+        robot_positions = self.get_team_positions(team)
+        if robot_id not in robot_positions:
+            # assert(len(robot_positions) <= 6)
+            robot_positions[robot_id] = deque([], ROBOT_POS_HISTORY_LENGTH)
+        robot_positions[robot_id].appendleft((time.time(), pos))
+
+    def remove_robot(self, team, robot_id):
+        team_positions = self.get_team_positions(team)
+        del team_positions[robot_id]
+        team_commands = self.get_team_commands(team)
+        if robot_id in team_commands:
+            del team_commands[robot_id]
+        team_status = self.get_team_status(team)
+        if robot_id in team_status:
+            del team_status[robot_id]
+
+    def get_robot_last_update_time(self, team, robot_id):
+        robot_positions = self.get_team_positions(team)
+        if robot_id not in robot_positions:
+            # print("getting update time of robot never seen?!?")
+            return None
+        timestamp, pos = robot_positions[robot_id][0]
+        # remove lost robots after a while
+        if time.time() - timestamp > ROBOT_REMOVE_TIME:
+            self.remove_robot(team, robot_id)
+        return timestamp
+
+    # check if robot hasn't been seen by cameras in a while
+    def is_robot_lost(self, team, robot_id):
+        last_update_time = self.get_robot_last_update_time(team, robot_id)
+        if last_update_time is None:
+            return True
+        return time.time() - last_update_time > ROBOT_LOST_TIME
+
+    def get_team_commands(self, team):
+        if team == 'blue':
+            return self._blue_robot_commands
+        else:
+            assert(team == 'yellow')
+            return self._yellow_robot_commands
+
+    def get_team_status(self, team):
+        if team == 'blue':
+            return self._blue_robot_status
+        else:
+            assert(team == 'yellow')
+            return self._yellow_robot_status
+
+    def get_robot_commands(self, team, robot_id):
+        team_commands = self.get_team_commands(team)
+        if robot_id not in team_commands:
+            team_commands[robot_id] = RobotCommands()
+            team_commands[robot_id].logger = self.logger
+        return team_commands[robot_id]
+
+    def get_robot_status(self, team, robot_id):
+        team_status = self.get_team_status(team)
+        if robot_id not in team_status:
+            team_status[robot_id] = RobotStatus()
+            team_status[robot_id].logger = self.logger
```

Comment:
I would leave a comment explaining the logger set in these functions too! And also a note in the class __init__ functions where the logger is initialized to None, so people know that it actually gets set over here dynamically.

Also it may not work properly if unless you set the logger EVERY time, including when you are just returning team_commands[robot_id]. That's because the gamestate may be called from a different provider.